File: VB/Wpf-Dashboard-Custom-Properties/Modules/ItemDescriptionModule/ItemDescriptionModule.vb
Imports System
Imports System.Globalization
Imports System.Windows
Imports System.Windows.Data
Imports DevExpress.DashboardCommon
Imports DevExpress.DashboardWpf
Imports DevExpress.Mvvm.UI.Interactivity

Namespace Wpf_Dashboard_Custom_Properties

    Public Class ItemDescriptionModule
        Inherits Behavior(Of DashboardControl)

        Public Const ItemDescriptionPropertyName As String = "Description"

        Protected Overrides Sub OnAttached()
            MyBase.OnAttached()
            Dim resources As ResourceDictionary = CType(Application.LoadComponent(New Uri("/Wpf-Dashboard-Custom-Properties;component/Modules/ItemDescriptionModule/ItemResources.xaml", UriKind.Relative)), ResourceDictionary)
            AssociatedObject.DashboardItemStyle = TryCast(resources("itemStyle"), Style)
        End Sub

        Protected Overrides Sub OnDetaching()
            AssociatedObject.DashboardItemStyle = Nothing
            MyBase.OnDetaching()
        End Sub
    End Class

    Public Class ShowItemDescriptionCommandConverter
        Implements IValueConverter

        Public Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
            Return New RelayCommand(AddressOf ShowDescriptionCommand)
        End Function

        Private Sub ShowDescriptionCommand(ByVal value As Object)
            Dim dashboardItem As DashboardItem = CType(value, DashboardItem)
            MessageBox.Show(dashboardItem.CustomProperties(ItemDescriptionModule.ItemDescriptionPropertyName), "Dashboard Item Description")
        End Sub

        Public Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Throw New NotImplementedException()
        End Function
    End Class

    Public Class ShowItemDescriptionCommandParameterConverter
        Implements IMultiValueConverter

        Public Function Convert(ByVal values As Object(), ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IMultiValueConverter.Convert
            Dim itemName As String = CStr(values(0))
            Dim provider As IDashboardControlProvider = CType(values(1), IDashboardControlProvider)
            Return provider.Dashboard.Items(itemName)
        End Function

        Public Function ConvertBack(ByVal value As Object, ByVal targetTypes As Type(), ByVal parameter As Object, ByVal culture As CultureInfo) As Object() Implements IMultiValueConverter.ConvertBack
            Throw New NotImplementedException()
        End Function
    End Class

    Public Class ButtonVisibleConverter
        Implements IMultiValueConverter

        Public Function Convert(ByVal values As Object(), ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IMultiValueConverter.Convert
            Dim itemName As String = CStr(values(0))
            Dim provider As IDashboardControlProvider = CType(values(1), IDashboardControlProvider)
            Dim dashboardItem As DashboardItem = provider.Dashboard.Items(itemName)
            Return Not String.IsNullOrEmpty(dashboardItem.CustomProperties(ItemDescriptionModule.ItemDescriptionPropertyName))
        End Function

        Public Function ConvertBack(ByVal value As Object, ByVal targetTypes As Type(), ByVal parameter As Object, ByVal culture As CultureInfo) As Object() Implements IMultiValueConverter.ConvertBack
            Throw New NotImplementedException()
        End Function
    End Class
End Namespace

File: CS/Wpf-Dashboard-Custom-Properties/Modules/ItemDescriptionModule/ItemDescriptionModule.cs
using System;
using System.Globalization;
using System.Windows;
using System.Windows.Data;
using DevExpress.DashboardCommon;
using DevExpress.DashboardWpf;
using DevExpress.Mvvm.UI.Interactivity;

namespace Wpf_Dashboard_Custom_Properties {
    public class ItemDescriptionModule : Behavior<DashboardControl> {
        public const string ItemDescriptionPropertyName = "Description";

        protected override void OnAttached() {
            base.OnAttached();
            ResourceDictionary resources = (ResourceDictionary)Application.LoadComponent(new Uri("/Wpf-Dashboard-Custom-Properties;component/Modules/ItemDescriptionModule/ItemResources.xaml", UriKind.Relative));
            AssociatedObject.DashboardItemStyle = resources["itemStyle"] as Style;
        }
        protected override void OnDetaching() {
            AssociatedObject.DashboardItemStyle = null;
            base.OnDetaching();
        }
    }
    public class ShowItemDescriptionCommandConverter : IValueConverter {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture) {
            return new RelayCommand(ShowDescriptionCommand);
        }
        void ShowDescriptionCommand(object value) {
            DashboardItem dashboardItem = (DashboardItem)value;
            MessageBox.Show(dashboardItem.CustomProperties[ItemDescriptionModule.ItemDescriptionPropertyName], "Dashboard Item Description");
        }
        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }
    public class ShowItemDescriptionCommandParameterConverter : IMultiValueConverter {
        public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture) {
            string itemName = (string)values[0];
            IDashboardControlProvider provider = (IDashboardControlProvider)values[1];
            return provider.Dashboard.Items[itemName];
        }
        public object[] ConvertBack(object value, Type[] targetTypes, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }
    public class ButtonVisibleConverter : IMultiValueConverter {
        public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture) {
            string itemName = (string)values[0];
            IDashboardControlProvider provider = (IDashboardControlProvider)values[1];
            DashboardItem dashboardItem = provider.Dashboard.Items[itemName];
            return !string.IsNullOrEmpty(dashboardItem.CustomProperties[ItemDescriptionModule.ItemDescriptionPropertyName]);
        }
        public object[] ConvertBack(object value, Type[] targetTypes, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }
}